### PR TITLE
feat!: The `HasTappableComponents` mixin is no longer needed

### PR DESCRIPTION
**Pull request by spydon**
# Description

This PR is the second in a series of refactors that aim to simplify event handling in Flame. The approach is as follows:

- Added the `MultiTapDispatcher` component, which contains the logic that used to be within the `HasTappableComponents` mixin. This component is internal; it mounts to a `FlameGame` directly, and ensures that it is a singleton.
- Whenever any `TapCallbacks` component is added to a game, it automatically adds the `MultiTapDispatcher` component (unless there is already one), which in turn registers a tap gesture detector with `GestureDetectorBuilder` and rebuilds the game widget.

The end result is that now in order to make a component tappable you only need to add the `TapCallbacks` mixin to that component, everything else will be handled by the framework.

Consequently, the `HasTappableComponents` mixin is now empty and marked as deprecated.

## Checklist

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [x] Yes, this PR is a breaking change.

### Migration instructions

The `HasTappableComponents` mixin is now empty & deprecated. If your game used this mixin overriding its methods `onTapDown`, `onTapUp`, etc -- then they will no longer work. If you want to receive tap events at the top level of the game, then simply add a `TapCallbacks` component to the top level of the game.

## Related Issues

WIP for #1733

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md

### CODE REVIEW

File: examples/lib/stories/bridge_libraries/flame_isolate/simple_isolate_example.dart [hunk 1]
```diff
@@ -2,12 +2,13 @@ import 'dart:math';
 
 import 'package:flame/components.dart';
 import 'package:flame/effects.dart';
+import 'package:flame/experimental.dart';
```
Comment (luanpotter) [inline examples/lib/stories/bridge_libraries/flame_isolate/simple_isolate_example.dart:5]: why is this needed here?

File: packages/flame/test/game/mixins/has_tappables_test.dart [hunk 2]
```diff
@@ -132,9 +135,9 @@ void main() {
         await game.ready();
         game.onTapDown(
           1,
-          createTapDownEvent(
+          TapDownInfo.fromDetails(
             game,
```
[the remark this answers is not in the document]
Comment (spydon) [reply packages/flame/test/game/mixins/has_tappables_test.dart:140]: True, it was just left after iterations of moving things over, it's removed now.

File: examples/lib/stories/bridge_libraries/flame_isolate/simple_isolate_example.dart (hunk 1 above)
Comment (spydon) [reply examples/lib/stories/bridge_libraries/flame_isolate/simple_isolate_example.dart:5]: Because the `TapCallbacks` mixin is exposed from there, might be time to expose it elsewhere too.

File: doc/flame/inputs/tap_events.md [hunk 3]
```diff
@@ -1,8 +1,8 @@
 # Tap Events
 
 ```{note}
-This document describes a new experimental API. The more traditional approach
-for handling tap events is described in [](gesture_input.md).
+This document describes the new ta events API. The old (legacy) approach,
```
Comment (erickzanardo) [inline doc/flame/inputs/tap_events.md:4]:
```suggestion
This document describes the new tap events API. The old (legacy) approach,
```

File: doc/flame/inputs/tap_events.md [hunk 4]
```diff
@@ -15,33 +15,24 @@ Multiple tap events can occur at the same time, especially if the user has multi
...
 `pointerId` property.
 
-It takes only a few simple steps to enable these events for your game:
+For those components that you want to respond to taps, add the `TapCallbacks` mixin.
 
-1. Add the `HasTappableComponents` mixin to your main game class:
+- This mixin adds four overridable methods to your component: `onTapDown`, `onTapUp`,
+  `onTapCancel`, and `onLongTapDown`. By default, each of these methods does nothing, they need
+  to be overridden in order to perform any function.
+- In addition, the component must implement the `containsLocalPoint()` method -- this method
```
Comment (erickzanardo) [inline doc/flame/inputs/tap_events.md:23]: Can we have a default implementation for `containsLocalPoint`? Some that would at least rely on the bounding rectangle of the component?

Comment (erickzanardo) [reply doc/flame/inputs/tap_events.md:23]: Okay, reading down bellow, I see that this is the case for `PositionComponents`, then probably just change this to "unless your component is a `PositionComponent` the component must implement..."?

File: examples/lib/stories/bridge_libraries/flame_isolate/simple_isolate_example.dart (hunk 1 above)
Comment (erickzanardo) [reply examples/lib/stories/bridge_libraries/flame_isolate/simple_isolate_example.dart:5]: I would say so, yeah, since we are making it general available, and the older API is now deprecated, I would argue that this is not experimental anymore.

File: packages/flame/example/lib/main.dart [hunk 5]
```diff
@@ -17,20 +17,20 @@ void main() {
...
 
-  @override
-  void onTapDown(TapDownEvent event) {
-    super.onTapDown(event);
-    if (!event.handled) {
-      final touchPoint = event.canvasPosition;
-      add(Square(touchPoint));
-    }
-  }
+  //@override
```
Comment (erickzanardo) [inline packages/flame/example/lib/main.dart:26]: Maybe this was forgotten?

File: packages/flame/test/events/component_mixins/drag_callbacks_test.dart [hunk 6]
```diff
@@ -131,10 +132,38 @@ void main() {
...
+
+    testWidgets(
+      'drag correctly registered handled event directly on FlameGame',
+      (tester) async {
+        final game = _DragCallbacksGame()..onGameResize(Vector2.all(300));
+        await tester.pumpWidget(GameWidget(game: game));
+        await tester.pump();
+        await tester.pump();
+        expect(game.children.length, 1);
+        expect(game.isMounted, true);
```
Comment (erickzanardo) [inline packages/flame/test/events/component_mixins/drag_callbacks_test.dart:154]:
```suggestion
        expect(game.children.length, equals(1));
        expect(game.isMounted, isTrue);
```

Comment (erickzanardo) [reply packages/flame/test/events/component_mixins/drag_callbacks_test.dart:154]: The same suggestion to anywhere where a similar change would apply :)

**Review by erickzanardo — COMMENTED** [review]
Great job 🔥! Overall LGTM, just left a couple of simple comments!

File: examples/lib/stories/bridge_libraries/flame_isolate/simple_isolate_example.dart (hunk 1 above)
Comment (spydon) [reply examples/lib/stories/bridge_libraries/flame_isolate/simple_isolate_example.dart:5]: The older API is not deprecated yet, it is only the `HasTappableComponents` mixin that is deprecated in this PR, and that mixin is part of the new API. I would keep it as experimental for one more version to make sure that we don't want to make any more breaking changes to it.

Comment (erickzanardo) [reply examples/lib/stories/bridge_libraries/flame_isolate/simple_isolate_example.dart:5]: Got ya, so the older api can be used yet, you just don't need the mixin on the game class, sounds good then!!!

Comment (spydon) [reply examples/lib/stories/bridge_libraries/flame_isolate/simple_isolate_example.dart:5]: You almost got it, the older API is using a mixin called `HasTappables` and that one has to still be used if one wants to use the `Tappable` API instead of the `TapCallbacks` API.

File: doc/flame/inputs/tap_events.md (hunk 4 above)
Comment (spydon) [reply doc/flame/inputs/tap_events.md]: Good idea, updated!

File: packages/flame/example/lib/main.dart (hunk 5 above)
Comment (spydon) [reply packages/flame/example/lib/main.dart:26]: Oooops, good catch!

File: packages/flame/test/events/component_mixins/drag_callbacks_test.dart (hunk 6 above)
Comment (spydon) [reply packages/flame/test/events/component_mixins/drag_callbacks_test.dart]: Fixed! We have a loooot of other tests to fix this in too.

**Review by erickzanardo — APPROVED** [review]
LGTM awesome job!!!
